client/extensions/woocommerce/state/ui/shipping/zones/methods/selectors.js

```diff
@@ -0,0 +1,159 @@
+import { mergeMethodEdits } from './helpers';
+
+const getShippingZone = ( state, zoneId, siteId ) => {
+	if ( ! areShippingZonesLoaded( state, siteId ) ) {
+		return null;
+	}
+	if ( null === zoneId ) {
+		return getCurrentlyEditingShippingZone( state, siteId );
+	}
+	return find( getAPIShippingZones( state, siteId ), { id: zoneId } );
+};
+
+const getShippingZoneMethodsEdits = ( state, zoneId, siteId ) => {
+	const zonesEdits = getShippingZonesEdits( state, siteId );
+	if ( zonesEdits ) {
+		const zoneEdits = find( zonesEdits[ getBucket( { id: zoneId } ) ], { id: zoneId } );
+		if ( zoneEdits && ! isEmpty( zoneEdits.methods ) ) {
+			return zoneEdits.methods;
+		}
+	}
+	return {
+		creates: [],
+		updates: [],
+		deletes: [],
+	};
+};
+
+const sortShippingZoneMethods = ( state, siteId, methods ) => {
+	return methods.sort( ( a, b ) => {
+		const aId = isNil( a._originalId ) ? a.id : a._originalId;
+		const bId = isNil( b._originalId ) ? b.id : b._originalId;
+
+		if ( isNumber( aId ) ) {
+			// Both IDs are numbers (come from the server), so compare their "order" property
+			if ( isNumber( bId ) ) {
+				return getShippingZoneMethod( state, aId, siteId ).order - getShippingZoneMethod( state, bId, siteId ).order;
+			}
+			// "a" is a pre-existing method (numeric ID) so it comes first than the newly created (object ID) "b"
+			return -1;
+		}
+
+		if ( isNumber( bId ) ) {
+			// "b" is a pre-existing method (numeric ID) so it comes first than the newly created (object ID) "a"
+			return 1;
+		}
+		// Both IDs are "Object IDs", just compare their indices
+		return aId.index - bId.index;
+	} );
+};
+
+const overlayShippingZoneMethods = ( state, zone, siteId, extraEdits ) => {
+	const methodIds = [ ...zone.methodIds ];
+	const edits = getShippingZoneMethodsEdits( state, zone.id, siteId );
+	const { creates, updates, deletes } = mergeMethodEdits( edits, extraEdits );
+
+	// Overlay the current edits on top of (a copy of) the wc-api zone methods
+	pullAll( methodIds, map( deletes, 'id' ) );
+	const methods = methodIds.map( methodId => getShippingZoneMethod( state, methodId, siteId ) );
+	updates.forEach( ( update ) => {
+		const index = methodIds.indexOf( update.id );
+		if ( -1 === index ) {
+			return;
+		}
+		methods[ index ] = { ...methods[ index ], ...update };
+	} );
+	return sortShippingZoneMethods( state, siteId, [ ...methods, ...creates ] );
+};
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Number} [zoneId] Shipping Zone ID
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {Array} The list of shipping methods included in the given shipping zone. On any failure, it will return
+ * an empty Array
+ */
+export const getShippingZoneMethods = ( state, zoneId, siteId = getSelectedSiteId( state ) ) => {
+	if ( ! areShippingZonesLoaded( state, siteId ) ) {
+		return [];
+	}
+	const zone = getShippingZone( state, zoneId, siteId );
```

Comment:
Fixed in https://github.com/Automattic/wp-calypso/pull/14890/commits/411893092b25286d1a6014c9b285f2c7253bd95d